src/crypto/CHIPCryptoPALmbedTLS.cpp

```diff
@@ -817,16 +993,76 @@ P256Keypair::~P256Keypair()
 
 CHIP_ERROR P256Keypair::NewCertificateSigningRequest(uint8_t * out_csr, size_t & csr_length) const
 {
+#if defined(MBEDTLS_USE_TINYCRYPT)
     CHIP_ERROR error = CHIP_NO_ERROR;
-    int result       = 0;
+
+    int result = 0;
+
+    size_t out_length;
+
+    const mbedtls_uecc_keypair * keypair = to_const_keypair(&mKeypair);
+
+    mbedtls_x509write_csr csr;
+    mbedtls_x509write_csr_init(&csr);
+
+    mbedtls_pk_context pk;
+    mbedtls_pk_init(&pk);
+
+    const mbedtls_pk_info_t * pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_ECKEY);
+    VerifyOrExit(pk_info != nullptr, error = CHIP_ERROR_INTERNAL);
+
+    VerifyOrExit(mInitialized, error = CHIP_ERROR_INCORRECT_STATE);
+
+    result = mbedtls_pk_setup(&pk, pk_info);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    memcpy(mbedtls_pk_uecc(pk), keypair, sizeof(mbedtls_uecc_keypair));
+
+    mbedtls_x509write_csr_set_key(&csr, &pk);
+
+    mbedtls_x509write_csr_set_md_alg(&csr, MBEDTLS_MD_SHA256);
```

Comment:
The `mbedtls_x509write_csr` parts are actually the same and should not be duplicated in both if/else